Feature/global search

AshTNA opened the pull request:
Hi @gtvj 

Here is the work-in-progress component for our global search.

Thanks

AshTNA commented:
Just a quick note - I am aware there is an issue with the search options box CSS, it doesn't stay the same width as it is resized, I am not sure how to fix it at the moment.

gtvj commented:
Hi @AshTNA - could you rebase to resolve conflicts?

gtvj commented on components/global_search/child_components/search_option/SearchOption.js:
```diff
@@ -0,0 +1,25 @@
+    render_search_options(arr) {
+        return arr.map((item, index) =>
+            <div key={index}>
+                <input name={this.props.group_name} type="radio" id={item.id}
```
Hi @AshTNA - @punalsc suggested that you could destructure `props` here (if you wanted to).

gtvj commented:
> Just a quick note - I am aware there is an issue with the search options box CSS, it doesn't stay the same width as it is resized, and I am not sure how to fix it at the moment.

I've had a look at this and think it might be that this is easiest to resolve if we move to specifying the width of _both_ the search box (`name="_q"`) and the individual search options (`label` elements) to share the same `rem` value (with a fallback to `em`) rather than using `%`. We'd also need to introduce any _necessary_ media queries (it might be we don't need any). To achieve this in dev tools I had to override a `max-width` and `width` on some containing elements.

gtvj commented:
Another issue I can see is that when the placeholder text is "Search Discovery, our catalogue" the field isn't big enough to accommodate it. Should the placeholders (and associated `aria-label` attributes) be changed to "Search our website" and "Search our records"?

gtvj commented on components/global_search/global-search.react.js:
```diff
@@ -0,0 +1,146 @@
+                select_type: 'Select a search type',
+                options: [
+                    {
+                        label: 'Search the website',
```
This should be 'Search our website'

gtvj commented on components/global_search/global-search.react.js:
```diff
@@ -0,0 +1,146 @@
+                        url: 'http://www.nationalarchives.gov.uk/search/results'
+                    },
+                    {
+                        label: 'Search Discovery, our catalogue',
```
This should be 'Search our records'

punalsc commented on components/global_search/global-search.react.js:
```diff
@@ -0,0 +1,146 @@
+    constructor(props) {
+        super(props);
+
+        this.handle_search_selection = this.handle_search_selection.bind(this);
```
@AshTNA , you could possibly cut your code down a lot, if you use `arrow functions`.
This way you don't need to bind your methods.

punalsc commented on components/global_search/global-search.react.js:
```diff
@@ -0,0 +1,146 @@
+        this.show_search_options_ref = React.createRef();
+        this.show_hide_mobile_ref = React.createRef();
+
+        this.state = {
```
You could possibly remove this `this` keyword. It will still work.

AshTNA replied:
Me and @punalsc sat down and had issues implementing this, so we will leave it as is for now.

AshTNA commented:
Hi @gtvj,

I have made most of the changes, except for the CSS one. I will probably need to pair up with a more senior developer to resolve that issue.

Now that I have hidden the search options on page load, the tests are failing as the search options box does not exist unless the arrow icon is interacted with. I am trying to look into how to fix it, though I may need some help with that as well.

Thanks :+1:

AshTNA commented:
As discussed, I will refactor the jest tests to Cypress when I have time, as jest is harder to work with when creating React components.

AshTNA commented:
@gtvj @punalsc I think all the changes have now been implemented. I have written some Cypress tests which you may want to check. The HTML seems to be passing the W3 Validator fine as well.

Thanks :+1:

punalsc commented:
> @gtvj @punalsc I think all the changes have now been implemented. I have written some Cypress tests which you may want to check. The HTML seems to be passing the W3 Validator fine as well.
> 
> Thanks 👍

Hi @AshTNA ,

Are we planning to use this component in for both website and discovery?

punalsc commented on components/global_search/global-search.react.js:
```diff
@@ -0,0 +1,162 @@
+
+    handle_search_selection = (e) => {
+        if (e.target.type === 'radio') {
+            let selection = this.state.search_options.options.find((i) => {
```
You don't need to use `this.state` as you are destructuring `this` on line `47`

punalsc commented on components/global_search/global-search.react.js:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    toggle_search_options = () => {
+        this.setState({show_search_options: !this.state.show_search_options})
```
Again same as above
You don't need to use `this.state` as you are destructuring `this` on line `47`

punalsc commented:
> > @gtvj @punalsc I think all the changes have now been implemented. I have written some Cypress tests which you may want to check. The HTML seems to be passing the W3 Validator fine as well.
> > Thanks 👍
> 
> Hi @AshTNA ,
> 
> Are we planning to use this component in for both website and discovery? The reason I ask this you might need to update the versioning for both website and discovery.

I have had a look at the PR and all seems fine, and happy to merge once we have fixed the merge conflict and versioning.

punalsc commented on webpack.config.js:
```diff
@@ -4,41 +4,52 @@ const discovery_version = `1.1.8`;
 const website_version = `1.0.0`;
```
Version number would need to be updated.

AshTNA commented:
Thanks @punalsc, I updated the version numbers to `1.1.0` and `1.2.0`. :+1:

Today I modified the CSS a bit to make it work on tna-base. The component does not look right on TNA or Discovery though which we'll need to fix, as you can see here:

- TNA: https://tna-global-search.netlify.com/tna.html
- Discovery: https://tna-global-search.netlify.com/discovery.html
- TNA-BASE (which is fine): https://tna-global-search.netlify.com

I am not sure about merging yet, @gtvj what do you think? For next steps I will:

- Decide on the SVG icons we use, as the ones we are using currently require attribution - maybe we can add the attribution to one page on the site somewhere
- Run the search by the UX team (do they need to do user testing?)
- Run the search by the Editorial team.

AshTNA commented:
I have spoken to the editorial team and they have suggest some changes that have been added to this PR.

I have also spoken to the UX team and they have identified some issues that they may want to address. As @gtvj noted, this component was designed to fix a technical accessibility issue, and therefore we will try and get the current iteration live as soon as possible. Any UX changes will be incorporated in a future pull request.

I have also spoken to @sdwilkes about the issue with SVG attribution, and we are going to try and see if I, or one of the designers can create some SVG's for us to use.

Therefore, the current outstanding tasks are:

- Fix the CSS issues on TNA
- Fix the CSS issues on Discovery
- Incorporate SVGs that we are allowed to use

AshTNA commented:
Note: when deploying this to TNA, TNA-Base, and Discovery, we need to remove the Javascript that currently shows the search expander icon on mobile, as it is being injected into the page on load via a JS file

AshTNA commented:
Hi all,

I was struggling with a CSS issue with Global Search - the search appeared off by a few pixels on each section of our website (TNA, TNA-Base, Discovery). @netdesignr helped me with this issue and he discovered that this is because the initial border widths on each section of our site are different. Therefore when I am applying borders on elements, some are having their default replaced and therefore remaining the same width, and others are having a border added to themselves and therefore their widths were increasing. This resulted in slight pixel differences on each section of the site.

Therefore, we have decided that it might best to pass a prop through in the HTML to tell the component which section of the site it is on. The CSS would then adjust according to this prop. We will probably have to do this for SSO and MYC as well.

@gtvj, what do you think of this approach? I may be describing the CSS issue incorrectly, so feel free to speak to @netdesignr about this.

Thanks

gtvj commented:
Hi @AshTNA - I can see this approach would work but would only want to consider it as a last resort. I say this because it kind of goes against the spirit of components (which should have absolute control over their appearance, achieved via overrides where necessary). 

I'm surprised that we're not able to override the issue within the component. I said this to @netdesignr also on Friday when he explained it to me. 

@AshTNA - could you pop some time in the calendar for you and I to look at this and, if we're not able to fix it, I'm happy with the props approach described.

AshTNA commented:
Hi all,

As the current component has more complexity than necessary, and contained a odd bug, it has been refactored by @netdesignr and therefore the new code resides in https://github.com/nationalarchives/tna-components/tree/feature/newGlobalSearch

AshTNA commented:
This now exists at https://github.com/nationalarchives/tna-components/tree/feature/global-search
